In [52]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error , r2_score
import joblib
import os
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error,r2_score
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

In [53]:
df=pd.read_csv(r'C:\Users\91948\CondaIBM\Final Project\Data Colletion\Wind turbine.csv')
df


Date/Time  LV ActivePower (kW)  Wind Speed (m/s)  \
0      01 01 2018 00:00           380.047791          5.311336   
1      01 01 2018 00:10           453.769196          5.672167   
2      01 01 2018 00:20           306.376587          5.216037   
3      01 01 2018 00:30           419.645905          5.659674   
4      01 01 2018 00:40           380.650696          5.577941   
...                 ...                  ...               ...   
50525  31 12 2018 23:10          2963.980957         11.404030   
50526  31 12 2018 23:20          1684.353027          7.332648   
50527  31 12 2018 23:30          2201.106934          8.435358   
50528  31 12 2018 23:40          2515.694092          9.421366   
50529  31 12 2018 23:50          2820.466064          9.979332   

       Theoretical_Power_Curve (KWh)  Wind Direction (°)  
0                         416.328908          259.994904  
1                         519.917511          268.641113  
2                         390.900016          272.564789  
3                         516.127569          271.258087  
4                         491.702972          265.674286  
...                              ...                 ...  
50525                    3397.190793           80.502724  
50526                    1173.055771           84.062599  
50527                    1788.284755           84.742500  
50528                    2418.382503           84.297913  
50529                    2779.184096           82.274620  

[50530 rows x 5 columns]

In [54]:
df.shape

(50530, 5)

In [55]:
df.mode()

Date/Time  LV ActivePower (kW)  Wind Speed (m/s)  \
0      01 01 2018 00:00                  0.0               0.0   
1      01 01 2018 00:10                  NaN               NaN   
2      01 01 2018 00:20                  NaN               NaN   
3      01 01 2018 00:30                  NaN               NaN   
4      01 01 2018 00:40                  NaN               NaN   
...                 ...                  ...               ...   
50525  31 12 2018 23:10                  NaN               NaN   
50526  31 12 2018 23:20                  NaN               NaN   
50527  31 12 2018 23:30                  NaN               NaN   
50528  31 12 2018 23:40                  NaN               NaN   
50529  31 12 2018 23:50                  NaN               NaN   

       Theoretical_Power_Curve (KWh)  Wind Direction (°)  
0                                0.0                 0.0  
1                                NaN                 NaN  
2                                NaN                 NaN  
3                                NaN                 NaN  
4                                NaN                 NaN  
...                              ...                 ...  
50525                            NaN                 NaN  
50526                            NaN                 NaN  
50527                            NaN                 NaN  
50528                            NaN                 NaN  
50529                            NaN                 NaN  

[50530 rows x 5 columns]

In [56]:
df.mode()

Date/Time  LV ActivePower (kW)  Wind Speed (m/s)  \
0      01 01 2018 00:00                  0.0               0.0   
1      01 01 2018 00:10                  NaN               NaN   
2      01 01 2018 00:20                  NaN               NaN   
3      01 01 2018 00:30                  NaN               NaN   
4      01 01 2018 00:40                  NaN               NaN   
...                 ...                  ...               ...   
50525  31 12 2018 23:10                  NaN               NaN   
50526  31 12 2018 23:20                  NaN               NaN   
50527  31 12 2018 23:30                  NaN               NaN   
50528  31 12 2018 23:40                  NaN               NaN   
50529  31 12 2018 23:50                  NaN               NaN   

       Theoretical_Power_Curve (KWh)  Wind Direction (°)  
0                                0.0                 0.0  
1                                NaN                 NaN  
2                                NaN                 NaN  
3                                NaN                 NaN  
4                                NaN                 NaN  
...                              ...                 ...  
50525                            NaN                 NaN  
50526                            NaN                 NaN  
50527                            NaN                 NaN  
50528                            NaN                 NaN  
50529                            NaN                 NaN  

[50530 rows x 5 columns]

In [57]:
df.describe().T

count         mean          std       min  \
LV ActivePower (kW)            50530.0  1307.684332  1312.459242 -2.471405   
Wind Speed (m/s)               50530.0     7.557952     4.227166  0.000000   
Theoretical_Power_Curve (KWh)  50530.0  1492.175463  1368.018238  0.000000   
Wind Direction (°)             50530.0   123.687559    93.443736  0.000000   

                                      25%          50%          75%  \
LV ActivePower (kW)             50.677890   825.838074  2482.507568   
Wind Speed (m/s)                 4.201395     7.104594    10.300020   
Theoretical_Power_Curve (KWh)  161.328167  1063.776283  2964.972462   
Wind Direction (°)              49.315437    73.712978   201.696720   

                                       max  
LV ActivePower (kW)            3618.732910  
Wind Speed (m/s)                 25.206011  
Theoretical_Power_Curve (KWh)  3600.000000  
Wind Direction (°)              359.997589

In [58]:
df.isna().any()

Date/Time                        False
LV ActivePower (kW)              False
Wind Speed (m/s)                 False
Theoretical_Power_Curve (KWh)    False
Wind Direction (°)               False
dtype: bool

In [59]:
df.var()

C:\Users\91948\AppData\Local\Temp\ipykernel_15876\1568254755.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df.var()


LV ActivePower (kW)              1.722549e+06
Wind Speed (m/s)                 1.786893e+01
Theoretical_Power_Curve (KWh)    1.871474e+06
Wind Direction (°)               8.731732e+03
dtype: float64

In [60]:
x=df[['Wind Speed (m/s)']]
y = df['LV ActivePower (kW)']
y.head()

0    380.047791
1    453.769196
2    306.376587
3    419.645905
4    380.650696
Name: LV ActivePower (kW), dtype: float64

In [61]:
x.head()

Wind Speed (m/s)
0          5.311336
1          5.672167
2          5.216037
3          5.659674
4          5.577941

In [62]:
x_train, x_test,y_train, y_test = train_test_split(x,y ,
                                   random_state=94, 
                                   test_size=0.20)

In [63]:
x_train

Wind Speed (m/s)
33219          7.303417
3655           5.868001
24597         11.558850
4895           9.873700
8628          15.437600
...                 ...
26802          5.086676
50499          4.994785
34149          9.199881
6482          10.838520
43707          7.488448

[40424 rows x 1 columns]

In [64]:
y_train

33219     870.955688
3655        0.000000
24597    3054.933105
4895     2413.622070
8628     3476.139893
            ...     
26802     348.389313
50499     471.419708
34149    1994.834961
6482     3143.066895
43707       0.000000
Name: LV ActivePower (kW), Length: 40424, dtype: float64

In [65]:
x_test

Wind Speed (m/s)
25113          6.523438
29042          5.636841
5676          15.398290
47493          5.444841
11948          9.373304
...                 ...
20607          9.153069
22234          2.211192
8105          17.583490
28472          2.486562
2098           5.197076

[10106 rows x 1 columns]

In [66]:
y_test

25113     728.294678
29042     362.814514
5676     3460.604980
47493     488.096710
11948    2539.521973
            ...     
20607    2013.511963
22234       0.000000
8105     3461.856934
28472       0.000000
2098      304.907501
Name: LV ActivePower (kW), Length: 10106, dtype: float64

--Model training and testing

In [67]:
RFR= RandomForestRegressor(n_estimators = 750, max_depth = 4, max_leaf_nodes = 500, random_state = 1)

In [68]:
RFR.fit(x_train,y_train) 

RandomForestRegressor(max_depth=4, max_leaf_nodes=500, n_estimators=750,
                      random_state=1)

In [69]:
y_pred=RFR.predict(x_test)

In [48]:
y_pred

array([6.79387765e+02, 4.05061637e+02, 3.41751091e+03, ...,
       3.53480917e+03, 3.14414437e+00, 3.01472402e+02])

In [71]:
predicted=RFR.predict(x_train)

In [72]:
predicted

array([1001.05403685,  537.60142348, 3045.0687722 , ..., 1796.31561314,
       2636.33799561, 1002.07902887])

--For finding the accuracy

In [74]:
accuracy=r2_score(y_test,y_pred)

In [75]:
accuracy

0.910887548174758

--There the Accuracy value calculated :0.910887548174758.

--This is for saving the model

In [81]:
joblib.dump(RFR, "Output Power Predicted.sav")

['Output Power Predicted.sav']